libraries/plugins/tags/include/scorum/tags/tags_api_impl.hpp

```diff
@@ -1043,6 +1067,70 @@ class tags_api_impl
         }
         return result;
     }
+
+    template <typename Index> class index_traverse
+    {
+        typedef comment_index::index_iterator<by_parent>::type search_iterator;
+
+    public:
+        index_traverse(const Index& index)
+            : _index(index)
+        {
+        }
+
+        template <typename OnItem>
+        void find_comments(const std::string& parent_author, const std::string& parent_permlink, OnItem&& on_item)
+        {
+            account_name_type account_name = account_name_type(parent_author);
+
+            scan_children(account_name, parent_permlink);
+
+            while (!_stack.empty())
+            {
+                search_iterator itr = _stack.top();
+                _stack.pop();
+
+                on_item(*itr);
+
+                scan_children(itr->author, fc::to_string(itr->permlink));
+            }
+        }
+
+    private:
+        const Index& _index;
+
+        void scan_children(const account_name_type& parent_author, const std::string& parent_permlink)
+        {
+            auto itr = _index.find(boost::make_tuple(parent_author, parent_permlink));
```

Comment:
You can implement scan_children in more shorter way:
auto range = get_range_by<by_parent>(boost::make_tuple(parent_author, parent_permlink, 0) <= ::boost::lambda::_1,
                                             ::boost::lambda::_1 <= std::make_tuple(parent_author, parent_permlink, ALL_IDS));
std::reverse_copy (range.begin(), range.end(), _stack.begin());

Reply:
Good idea! It totally removes put_in_stack and change value type for std::stack from search_iterator to std::reference_wrapper<const object_type>. Look on_item(*itr). 
Just add method get_children to comment_service_i